## Santander product recommendation

### 모든 칼럼들에 대한 value 및 EDA 전처리, EDA 방향 잡기

About 'Primary customer'  
http://www.rhythmsystems.com/blog/what-is-the-difference-between-a-primary-and-secondary-core-customer

- fecha_dato(datetime) : 날짜 데이터(제대로 이해한게 맞다면 15-01-28 부터 16-05-28까지
- ncodpers(id) : 고객 번호(고유값?)
- ind_empleado(status_emplo) : 직원 상태
> A : 재직중
> B : 퇴사
> F : fillial(자식의?)
> N : 직원이 아님
> P : pasive?
- pais_residencia(residence) : 고객 거주지
- sexo(sex) : 고객 성별
- age(age) : 고객 나이
- fecha_alta(become_fh) : 고객이 은행에 first holder가 된 날
- ind_nuevo(new_or_not) : 새 고객인가 아닌가, 6개월 이내 등록했으면 새 고객
- antiguedad(seniority) : 고객 서열(월단위)
- indrel(pri_or_not) : 
> 1은 primary customer(구매의사가 충분한 고객?) 
> 99는 primary customer이다가 월말에는 primary customer가 아니게 됨
- ult_fec_cli_1t : primary customer 로써 마지막 date
- indrel_1mes(type_of_cus) : 월초의 고객 타입 
> 1 : Primary customer
> 2 : 공동 소유자
> P : 잠재 고객
> 3 : 이전 Primary 
> 4 : 이전 공동 소유자
- tiprel_1mes(relation_of_cus) :월초의 고객 관계
A : active
I : inactive
P : former customer
R : Potential
- indresi(resi_ind) : 
> S(Yes) or N(NO): 고객 거주지와 은행 위치가 **같은** 나라인가
- indext(foreign_ind) : 
> S(Yes) or N(NO) : 외국인의 출생국가와 은행 위치가 **다른가**?
- conyuemp(spouse) : 직원의 배우자 인가 아닌가 
> 1일 경우 직원의 배우자
- canal_entrada(?) : 고객이 조인하기 위해 사용되어진 채널?
- indfall : 죽은 고객?
- tipodom : 주소 타입
- cod_prov :  고객 주소(Province code)
- nomprov : province name
- ind_actividad_clie
> 1 : active customer
> 0 : inactive customer
- renta : gross income of the household(가정의 총 수입)
- segmento :
> 1 - VIP
> 2 - individuals
> 3 - college graduated
- ind_ahor_fin_ult1	Saving Account
- ind_aval_fin_ult1	Guarantees
- ind_cco_fin_ult1	Current Accounts
- ind_cder_fin_ult1	Derivada Account
- ind_cno_fin_ult1	Payroll Account
- ind_ctju_fin_ult1	Junior Account
- ind_ctma_fin_ult1	Más particular Account
- ind_ctop_fin_ult1	particular Account
- ind_ctpp_fin_ult1	particular Plus Account
- ind_deco_fin_ult1	Short-term deposits
- ind_deme_fin_ult1	Medium-term deposits
- ind_dela_fin_ult1	Long-term deposits
- ind_ecue_fin_ult1	e-account
- ind_fond_fin_ult1	Funds
- ind_hip_fin_ult1	Mortgage
- ind_plan_fin_ult1	Pensions
- ind_pres_fin_ult1	Loans
- ind_reca_fin_ult1	Taxes
- ind_tjcr_fin_ult1	Credit Card
- ind_valo_fin_ult1	Securities
- ind_viv_fin_ult1	Home Account
- ind_nomina_ult1	Payroll
- ind_nom_pens_ult1	Pensions
- ind_recibo_ult1	Direct Debit

- 스페인어로 된 columns name들을 영어로 바꾸는 건 어떨까

In [1]:
import pandas as pd
import seaborn as sns
import scipy as sp
import sklearn as sk
import matplotlib as mpl
mpl.use('Agg')
import matplotlib.pylab as plt
from mpl_toolkits.mplot3d import Axes3D


import numpy as np
import scipy as sp
import sklearn as sk

mpl.use('Agg')

%matplotlib inline

sns.set()

C:\Users\ledes\Anaconda3\lib\site-packages\ipykernel_launcher.py:6: UserWarning: 
This call to matplotlib.use() has no effect because the backend has already
been chosen; matplotlib.use() must be called *before* pylab, matplotlib.pyplot,
or matplotlib.backends is imported for the first time.

The backend was *originally* set to 'module://ipykernel.pylab.backend_inline' by the following code:
  File "C:\Users\ledes\Anaconda3\lib\runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "C:\Users\ledes\Anaconda3\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "C:\Users\ledes\Anaconda3\lib\site-packages\ipykernel_launcher.py", line 16, in <module>
    app.launch_new_instance()
  File "C:\Users\ledes\Anaconda3\lib\site-packages\traitlets\config\application.py", line 658, in launch_instance
    app.start()
  File "C:\Users\ledes\Anaconda3\lib\site-packages\ipykernel\kernelapp.py", line 477, in start
    ioloop.IOLoop.instance().start()
  File "C:\U

In [2]:
#데이터 셋 불러오기
train = pd.read_csv("train_ver2.csv")
test = pd.read_csv("test_ver2.csv")

C:\Users\ledes\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:2698: DtypeWarning: Columns (5,8,11,15) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)
C:\Users\ledes\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:2698: DtypeWarning: Columns (15) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [3]:
#데이터 column, row 파악
print("shape of train data : ", train.shape)
print("shape of test data : ", test.shape)

shape of train data :  (13647309, 48)
shape of test data :  (929615, 24)


### 칼럼별 Null 파악

In [4]:
train.isnull().sum()

fecha_dato                      0
ncodpers                        0
ind_empleado                27734
pais_residencia             27734
sexo                        27804
age                             0
fecha_alta                  27734
ind_nuevo                   27734
antiguedad                      0
indrel                      27734
ult_fec_cli_1t           13622516
indrel_1mes                149781
tiprel_1mes                149781
indresi                     27734
indext                      27734
conyuemp                 13645501
canal_entrada              186126
indfall                     27734
tipodom                     27735
cod_prov                    93591
nomprov                     93591
ind_actividad_cliente       27734
renta                     2794375
segmento                   189368
ind_ahor_fin_ult1               0
ind_aval_fin_ult1               0
ind_cco_fin_ult1                0
ind_cder_fin_ult1               0
ind_cno_fin_ult1                0
ind_ctju_fin_u

데이터의 대부분이 null 값인 column = ** ult_fec_cli_1t, conyuemp**

### 칼럼별 데이터 파악
- 시각화가 가능한 EDA : 앞 100만개를 자른 부분을 통해 국소적인 부분만 판단
- 시각화가 불가능한  EDA : 전체 데이터를 이용

In [5]:
#전체 데이터 셋 이용한 데이터의 unique value  파악
try :
    for col in range(len(train.columns)) : 
        print("Number of unique value in {0}:{1}".format(train.columns[col],len(train[train.columns[col]].unique())))
except Exception as e:
    print(e)
    pass

Number of unique value in fecha_dato:17
Number of unique value in ncodpers:956645
Number of unique value in ind_empleado:6
Number of unique value in pais_residencia:119
Number of unique value in sexo:3
Number of unique value in age:235
Number of unique value in fecha_alta:6757
Number of unique value in ind_nuevo:3
Number of unique value in antiguedad:507
Number of unique value in indrel:3
Number of unique value in ult_fec_cli_1t:224
Number of unique value in indrel_1mes:14
Number of unique value in tiprel_1mes:6
Number of unique value in indresi:3
Number of unique value in indext:3
Number of unique value in conyuemp:3
Number of unique value in canal_entrada:163
Number of unique value in indfall:3
Number of unique value in tipodom:2
Number of unique value in cod_prov:53
Number of unique value in nomprov:53
Number of unique value in ind_actividad_cliente:3
Number of unique value in renta:520995
Number of unique value in segmento:4
Number of unique value in ind_ahor_fin_ult1:2
Number of u

## fecha_dato

In [6]:
train["fecha_dato"].value_counts()

2016-05-28    931453
2016-04-28    928274
2016-03-28    925076
2016-02-28    920904
2016-01-28    916269
2015-12-28    912021
2015-11-28    906109
2015-10-28    892251
2015-09-28    865440
2015-08-28    843201
2015-07-28    829817
2015-06-28    632110
2015-05-28    631957
2015-04-28    630367
2015-03-28    629209
2015-02-28    627394
2015-01-28    625457
Name: fecha_dato, dtype: int64

#### Comment
fecha_dato (date) : 
- 날짜 데이터, str 형태
- pandas의 datetime으로 활용가능
- ** 사실상 카테고리로 봐도 무방할 듯**

## ncodpers

In [7]:
train["ncodpers"].value_counts()

32752      17
1273696    17
1224536    17
1199940    17
1208128    17
1150780    17
1158968    17
1117996    17
1126184    17
1134372    17
1142560    17
1085212    17
1093400    17
1101588    17
1109776    17
1052428    17
1068804    17
1076992    17
511476     17
478692     17
429532     17
1216348    17
1298292    17
1298293    17
1281916    17
863403     17
871591     17
879779     17
797835     17
806023     17
           ..
1552552     1
1294656     1
1549574     1
1552012     1
1181550     1
1402308     1
1212249     1
750260      1
1441293     1
1549400     1
1477022     1
1544365     1
1516202     1
1552553     1
1420428     1
1542208     1
1526182     1
1548821     1
1023450     1
1363915     1
1550404     1
1550297     1
1468385     1
1437091     1
1534863     1
1549150     1
1543059     1
1535935     1
1551255     1
1489485     1
Name: ncodpers, Length: 956645, dtype: int64

In [8]:
(train["ncodpers"].value_counts()).max()

17

#### Comment
ncodpers(id) : 
- 고객 개개인의 id로 볼 수있는 는 unique한 값이 아닌 중복되는 값이 존재 -> EDA과정에서 살펴볼 필요가 있음.
-  최대 17개까지 중복
- int 형태

## ind_empleado

In [9]:
train["ind_empleado"].value_counts()

N    13610977
B        3566
F        2523
A        2492
S          17
Name: ind_empleado, dtype: int64

#### Comment
ind_empleado(emplo_status) : 
- 칼럼 값들의 대부분이 N을 가지고 있음
- str 형태인데 왜 int64라고 뜨지?
- 이정도 차이면 직원인지 아닌지에 따라 0과 1로 나누어도 되지 않을까

## pais_residencia

In [10]:
train["pais_residencia"].value_counts()

ES    13553710
FR        5161
AR        4835
DE        4625
GB        4605
US        3651
CO        3526
IT        2947
RO        2931
MX        2573
BR        2351
VE        2331
EC        2169
CH        1995
BE        1526
BO        1514
PY        1430
PT        1419
CL         989
PE         900
RU         769
CU         758
NL         757
SE         603
PL         599
CN         563
UY         510
UA         493
BG         476
AT         476
        ...   
GW          34
PH          34
CG          34
ET          34
OM          22
KH          17
LY          17
BZ          17
MM          17
SL          17
CF          17
AL          17
GH          17
LV          17
KZ          17
LB          17
TG          17
CD          17
TN          17
GI          17
GE          17
KW          17
GM          17
IS          17
ML          17
JM          11
ZW          11
DJ          11
BM           6
MT           2
Name: pais_residencia, Length: 118, dtype: int64

#### Comment
pais_residencia(nationality) :
- 국적, CountryCode를 나타내는 듯
- 대부분의 값이 ES, 이정도 차이면 스페인인지 아닌지에 따라 0과 1로 나누어도 되지 않을까
- str 형태인데 왜 int64라고 뜨지?


## sexo

In [11]:
train["sexo"].value_counts()

V    7424252
H    6195253
Name: sexo, dtype: int64

#### Comment
sexo(sex) :
- V : 여성, H : 남성 -> 여성의 수가 130만명가량 많은 편
- str형태인데 왜 int64라고 뜨지?


## age

In [12]:
train["age"].value_counts()

23     542682
24     527526
22     524620
21     500941
20     370847
25     345664
26     250641
 23    237202
43     231501
44     229359
42     227626
45     224073
41     221432
46     214320
 22    211694
 24    207259
40     205668
47     204715
27     203881
48     194513
39     186583
49     179384
 21    175047
28     174169
50     169287
38     166836
51     159013
37     154889
52     152009
29     149706
        ...  
107       439
103       419
  3       415
108       384
104       307
105       307
110       218
109       205
111       203
106       192
107       155
112       140
  2       130
113        93
108        72
109        60
116        51
111        49
112        48
110        43
115        43
115        39
113        24
114        22
117        14
116        12
163         8
127         8
164         3
126         3
Name: age, Length: 235, dtype: int64

In [13]:
print("age의 null 개수 : ", train["age"].isnull().sum())
print("age의 value가 ' NA' 인 것의 개수 : ", len(train[train["age"] == ' NA']))

age의 null 개수 :  0
age의 value가 ' NA' 인 것의 개수 :  27734


In [14]:
train["age"].isnull().sum()

0

In [15]:
def to_int(x) :
    if x == " NA" :
        result = None
    else :
        result = int(x)
        
    return result

In [16]:
#value가 ' NA'인것을 제외한 후 int로 형 변환
train["age"] = train["age"].apply(to_int)

In [17]:
print("age의 null 개수 : ", train["age"].isnull().sum())
print("나이가 가장 많은 고객 : ", train["age"].max())
print("나이가 가장 어린 고객 : ", train["age"].min())
print("100세 이상인 고객 : ", len(train[train["age"] >= 100]))
print("15세 미만인 고객 : ", len(train[train["age"] < 16]))

age의 null 개수 :  27734
나이가 가장 많은 고객 :  164.0
나이가 가장 어린 고객 :  2.0
100세 이상인 고객 :  15919
15세 미만인 고객 :  95084


#### Comment
age(age) :
- ' NA', null값으로는 인식되지 않지만, 값이 없는 수치들 발견  
- to_int 함수를 제작
> str 형태들을 int 형으로 형변환  
> null값이나 마찬 가지인 ' NA' 들을 null 값으로 만들어줌  

## fecha_alta

In [18]:
train["fecha_alta"].value_counts()

2014-07-28    57389
2014-10-03    54287
2014-08-04    45746
2013-10-14    40804
2013-08-03    33414
2012-08-06    26223
2014-07-21    26027
2012-08-03    25409
2012-08-02    25004
2013-10-11    24751
2013-08-10    23766
2012-08-01    22935
2014-09-17    22810
2014-08-06    22756
2014-09-24    22696
2014-07-30    22695
2012-08-09    22433
2013-07-22    22309
2012-08-07    22052
2012-07-24    21840
2014-09-23    21422
2012-07-30    20591
2013-10-10    20241
2013-10-28    20089
2013-10-09    19954
2012-10-01    19920
2012-07-31    19902
2012-08-10    19676
2015-07-25    19638
2012-07-19    19610
              ...  
2011-04-30       11
2005-07-31       11
2011-09-11       11
2014-05-11       11
2012-04-29       11
2010-10-09       11
2012-04-01       11
2014-06-29       11
2013-08-17       11
2014-08-10       11
2011-11-20       11
2013-03-09       11
2010-07-17       11
2012-07-14       11
2011-05-29       11
2014-04-27       11
2012-09-30       11
2013-05-11       11
2010-10-31       11


In [19]:
#최초거래일이 가장 빠른 고객의 날짜
pd.to_datetime(train["fecha_alta"]).min()

Timestamp('1995-01-16 00:00:00')

In [20]:
#최초거래일이 가장 최근인 고객의 날짜
pd.to_datetime(train["fecha_alta"]).max()

Timestamp('2016-05-31 00:00:00')

#### Comment
fecha_alta(become_cus) : 
- str형태의 날짜 데이터
- fecha_dato와 함께 두 번째 날짜 데이터

## ind_nuevo

In [21]:
train["ind_nuevo"].value_counts()

0.0    12808368
1.0      811207
Name: ind_nuevo, dtype: int64

#### Comment
ind_nuevo(new_before_6) :
- 0이 대부분 -> 대부분의 고객이 새 고객이 아님
- 실수형 데이터

## antiguedad

In [22]:
train['antiguedad'].value_counts()

0          134335
1          130681
12         130589
3          125789
2          125133
5          122246
     21    118544
4          117501
14         112959
     12    112571
6          111329
17         109770
16         109440
7          109156
15         108838
13         107487
24         104439
10         104434
18         104287
26         103378
     10    101731
25          99037
8           98171
27          96927
28          96575
23          96281
21          96251
19          96000
     33     95100
38          93861
            ...  
    238      2948
    215      2904
    228      2876
    203      2510
    236      2378
248          2271
    230      2258
    241      1980
    226      1967
      1      1898
    233      1863
    240      1854
249          1777
250          1512
    227      1356
    242      1148
251          1071
    239       872
    243       713
252           676
    246       507
    244       450
253           416
    245       360
254       

In [23]:
def func_for_antiguedad(x) :
    if x == '     NA' :
        result = None
    else :
        result = int(x)
        
    return result

In [24]:
train['antiguedad'] = train['antiguedad'].apply(func_for_antiguedad)

#### Comment
antiguedad(seniority) : 
- 공백이 있는 형태의 str 타입 
- (-)가 있음을 발견(우선 방치)
- ** numeric 전환 후, unique value 개수가 507 -> 259개로 감소('12'와  '     12' 같은 값들이 하나로 합쳐짐) **

## indrel

In [25]:
train["indrel"].value_counts()

1.0     13594782
99.0       24793
Name: indrel, dtype: int64

#### Comment
indrel(primary_or_not) : 
- 1 일 경우 월초 primary, 99 일 경우 월말에는 primary가 아님
- 대부분이 월초 primary 상태


## ult_fec_cli_1t

In [26]:
train["ult_fec_cli_1t"].value_counts()

2015-12-24    763
2015-12-28    521
2015-07-09    443
2015-07-06    405
2015-07-01    401
2015-07-21    391
2015-07-07    340
2015-07-17    330
2015-07-10    320
2015-07-28    311
2015-07-24    299
2015-07-20    295
2015-07-22    275
2015-07-15    273
2015-07-03    262
2015-07-29    255
2015-07-13    229
2015-07-30    224
2015-12-21    206
2015-07-23    206
2015-07-27    203
2015-07-14    196
2015-07-08    175
2015-12-17    172
2016-01-19    169
2016-01-05    167
2015-07-02    166
2015-12-16    158
2015-07-16    147
2015-12-18    139
             ... 
2016-02-25     65
2016-02-18     64
2015-12-07     64
2015-08-25     63
2016-04-20     63
2015-08-12     62
2016-04-22     62
2015-09-03     61
2015-08-21     61
2016-05-05     61
2015-08-24     61
2015-11-26     61
2015-08-07     59
2015-08-18     59
2015-08-20     59
2016-03-03     59
2016-04-07     58
2015-09-21     57
2016-04-14     57
2016-05-13     55
2016-03-07     55
2015-11-06     54
2015-08-14     51
2016-03-24     49
2016-03-23

#### Comment
ult_fec_cli_1t(last_primary) :
- 마지막으로 primary customer 였던 날짜
- 대부분이 null

## indrel_1mes

In [27]:
train.shape

(13647309, 48)

In [28]:
train["indrel_1mes"].value_counts()

1.0    7277607
1.0    4017746
1      2195328
3.0       1804
3.0       1409
3         1137
P          874
2          483
2.0        479
2.0        355
4.0        143
4.0         93
4           70
Name: indrel_1mes, dtype: int64

In [29]:
#같은 값을 가져야 하지만, 타입이 달라 데이터가 분산되어 나옴 -> 통일
# 'P'를 제외한 모든 값들을 int로 형 변환
def func_for_indrel(x) :
    number = "0123456789"
    if type(x) == str :
        if x[0] in number :
            result = int(float(x))
        else :
            result = x
    else :
        result = x
    
    return result

In [30]:
train["indrel_1mes"] = train["indrel_1mes"].apply(func_for_indrel)

#### Comment
indrel_1mes(type_of_cus) : 
- 같은 값이나, 서로 다른 타입으로 되어있던 값들을 하나의 값으로 통일 시켜줌
- 1이 압도적으로 많음


## tiprel_1mes

In [31]:
train["tiprel_1mes"].value_counts()

I    7304875
A    6187123
P       4656
R        870
N          4
Name: tiprel_1mes, dtype: int64

#### Comment 
tiprel_1mes(relation_of_cus) : 
- 카테고리 설명에 없는 4개의 N 값 발견

## indresi

In [32]:
train["indresi"].value_counts()

S    13553711
N       65864
Name: indresi, dtype: int64

In [33]:
train["indresi"].isnull().sum()

27734

#### Comment
indresi(same_country) : 
- 특이사항 없음

## indext

In [34]:
train["indext"].value_counts()

N    12974839
S      644736
Name: indext, dtype: int64

In [35]:
train["indext"].isnull().sum()

27734

#### Comment
- null 값이 27734로 동일한 칼럼이 많은 것을 발견

In [36]:
#row별 null  개수를 확인...하려했으나 어찌된 일인지 이것도 수행하지 못해서 우선 패스
null_row = train.apply(lambda x: len(train.columns) - x.count(), axis=1)

## conyuemp

In [37]:
train["conyuemp"].value_counts()

N    1791
S      17
Name: conyuemp, dtype: int64

#### Comment
conyuemp(spouse) : 
-  배우자 여부이나, 대부분이 null

## canal_entrada

In [38]:
train["canal_entrada"].value_counts()

KHE    4055270
KAT    3268209
KFC    3098360
KHQ     591039
KFA     409669
KHK     241084
KHM     183924
KHD     116891
KHN     116608
KAS      86221
RED      75607
KAG      74295
KAY      67350
KAA      66656
KAB      62381
KAE      50764
KCC      49308
KBZ      46446
KHL      45128
KFD      44461
KAI      37699
KEY      35146
KAW      34275
KAR      32686
KAZ      32186
KAF      30559
007      29063
013      27139
KCI      26546
KAH      24875
        ...   
004        210
KCQ        198
KDN        197
KCR        196
KDH        191
KCO        179
KEE        175
KCP        158
K00        152
KBE        147
KEQ        146
KAV        107
KFB        107
KCT        105
KBX        102
KBP         85
KEM         68
KFV         67
KCX         67
KBN         61
KHA         51
KGC         28
KGU         28
KDI         17
KDB         17
KGN         17
025         11
KDL         11
KHS          5
KHR          1
Name: canal_entrada, Length: 162, dtype: int64

In [39]:
train["canal_entrada"].isnull().sum()

186126

#### Comment
canal_entrada(channel) :
- 고객이 은행에 접촉하게 된 계기, 코드에 대한 이해는 아직 찾지 못함

## indfall

In [40]:
train["indfall"].value_counts()

N    13584813
S       34762
Name: indfall, dtype: int64

In [41]:
train["indfall"].isnull().sum()

27734

#### Comment
indfall(dead) : 
- 고객이 죽었나 살았나.
- 살아있는 사람이 대부분

## tipodom

In [42]:
train["tipodom"].value_counts()

1.0    13619574
Name: tipodom, dtype: int64

In [43]:
train["tipodom"].isnull().sum()

27735

#### Comment
tipodom(address) : 
- primary의 주소 타입, 전부 1
-  null값을 제외하고 전부 1, 어떻게 활용할까

## cod_prov

In [44]:
train["cod_prov"].value_counts()

28.0    4409600
8.0     1275219
46.0     682304
41.0     605164
15.0     429322
30.0     396759
29.0     367023
50.0     342543
3.0      313397
11.0     294684
36.0     280026
33.0     265749
47.0     238259
35.0     235700
6.0      192996
48.0     185888
45.0     183067
18.0     178562
37.0     164238
39.0     155706
14.0     144679
10.0     129912
7.0      124933
21.0     122283
13.0     119390
2.0      114128
12.0     104295
43.0     100115
9.0       97188
17.0      90538
31.0      88618
26.0      85202
27.0      84962
32.0      84009
24.0      83003
25.0      79059
20.0      71567
38.0      70968
19.0      64618
23.0      64363
4.0       60291
16.0      57310
49.0      50927
34.0      49282
40.0      42341
22.0      40181
5.0       38783
1.0       37704
44.0      22525
42.0      17660
52.0       9460
51.0       7218
Name: cod_prov, dtype: int64

![santander_prov_code](C:\Users\ledes\study\DNA_TeamProject\kang\santander_prov_code.jpg)

#### Comment
cod_prov(prov_code) :
- 스페인 각 지방의 고유 번호.
- 은행과의 거리를 점쳐볼 수 있을 듯(santander bank는 스페인 전국에 3개가 있는 곳으로 추정)

## nomprov

In [45]:
train["nomprov"].value_counts()

MADRID                    4409600
BARCELONA                 1275219
VALENCIA                   682304
SEVILLA                    605164
CORUÑA, A                  429322
MURCIA                     396759
MALAGA                     367023
ZARAGOZA                   342543
ALICANTE                   313397
CADIZ                      294684
PONTEVEDRA                 280026
ASTURIAS                   265749
VALLADOLID                 238259
PALMAS, LAS                235700
BADAJOZ                    192996
BIZKAIA                    185888
TOLEDO                     183067
GRANADA                    178562
SALAMANCA                  164238
CANTABRIA                  155706
CORDOBA                    144679
CACERES                    129912
BALEARS, ILLES             124933
HUELVA                     122283
CIUDAD REAL                119390
ALBACETE                   114128
CASTELLON                  104295
TARRAGONA                  100115
BURGOS                      97188
GIRONA        

#### Comment :
nomprov(name_prov)
- 스페인 각 지방 이름, 어떤 지방에 사는 고객인지
- prov_code와 중복

## ind_actividad_cliente

In [46]:
train["ind_actividad_cliente"].value_counts()

0.0    7384390
1.0    6235185
Name: ind_actividad_cliente, dtype: int64

In [47]:
train["ind_actividad_cliente"].isnull().sum()

27734

## renta

In [48]:
train["renta"].value_counts()

451931.22    5582
463625.16    1743
181042.20    1493
128318.52    1404
105260.88    1354
488798.49    1332
127141.50    1183
283325.67    1148
132335.73    1047
236690.34    1009
104563.80     954
227267.49     896
273387.54     886
111933.69     844
174407.10     823
227397.72     813
155930.43     804
218030.01     794
163432.47     712
208961.79     680
555605.55     677
168733.62     669
135522.15     657
148736.19     653
152601.99     649
248233.11     627
288997.44     625
326853.09     624
47550.39      607
91673.25      586
             ... 
31253.64        1
73367.82        1
46808.19        1
179561.43       1
108875.73       1
128959.86       1
39686.61        1
192297.42       1
74075.85        1
55041.57        1
65108.82        1
86106.78        1
144388.74       1
156827.52       1
80427.69        1
195859.59       1
97200.18        1
65845.32        1
130968.03       1
19559.37        1
147186.60       1
102417.57       1
79782.66        1
75676.77        1
238224.03 

#### Comment:
renta(income) : 
- 수입, 실수형인데 중복되는 데이터가 5000개씩 되는가?
- 아라님이 하신 EDA 부분을 봐야할 듯 

## segmento

In [49]:
train["segmento"].value_counts()

02 - PARTICULARES     7960220
03 - UNIVERSITARIO    4935579
01 - TOP               562142
Name: segmento, dtype: int64

In [50]:
#숫자와 텍스트가 함께 있는 형태를 숫자로만 변환
def func_for_segmento(x) :    
    if type(x) != float :
        if "1" in x :
            result = 1
        elif "2" in x :
            result = 2
        elif "3" in x :
            result = 3
    else : 
        result = None
    return result

In [51]:
train["segmento"] = train["segmento"].apply(func_for_segmento)

In [52]:
train["segmento"].value_counts()

2.0    7960220
3.0    4935579
1.0     562142
Name: segmento, dtype: int64

#### Comment 
segmento(segment) : 
- 고객을 3가지로 구분

## ind_ahor_fin_ult1

In [53]:
train["ind_ahor_fin_ult1"].value_counts()

0    13645913
1        1396
Name: ind_ahor_fin_ult1, dtype: int64

In [54]:
train["ind_aval_fin_ult1"].value_counts()

0    13646993
1         316
Name: ind_aval_fin_ult1, dtype: int64